packages/components/src/progressive-disclosure-panel/index.js

```diff
@@ -0,0 +1,140 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+import noop from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useState } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import ProgressiveDisclosurePanelTitle from './title';
+
+const ProgressiveDisclosurePanel = ( props ) => {
+	const { children, className, label: menuLabel, resetAll, title } = props;
+	const [ menuItems, setMenuItems ] = useState( {} );
+	const [ defaultChildren, setDefaultChildren ] = useState( {} );
+
+	// When conditionally including components e.g. { isShown && <Component /> }
+	// a boolean `false` will be passed as a child if component is excluded.
+	// This panel is only interested in the children to be displayed.
+	const filteredChildren = Array.isArray( children )
+		? children.filter( Boolean )
+		: [];
+
+	// Collect which children have custom values. Used to update menu state to
+	// reflect customization for children that display by default / always show.
+	const customizedChildren = filteredChildren.map( ( child ) =>
+		child.props.hasValue( child.props ) ? child.props.label : undefined
+	);
+
+	// On first render determine initial menu state and which children should
+	// always display by default.
+	useEffect( () => {
+		const items = {};
+		const defaults = {};
+
+		filteredChildren.forEach( ( child ) => {
+			items[ child.props.label ] = child.props.hasValue( child.props );
+			defaults[ child.props.label ] = child.props.isShownByDefault;
+		} );
+
+		setMenuItems( items );
+		setDefaultChildren( defaults );
+	}, [] );
```

Comment:
Thanks, this has been tightened up in 1fde44bb59.

I've removed the state for the default controls. Now whenever the panel's filtered children change, the `menuItems` state is updated. A hook now create a new menuItems object based on the current children. This will add any new children and exclude any removed children from the menu items. Previous selections of menu items are maintained.

Combining the above changes with your suggestion to add a ProgressiveDisclosurePanel.Item component also gives us greater flexibility when it comes to non-menu items being displayed within the panel.